worker/provisioner/provisioner_task.go

```diff
@@ -991,20 +975,26 @@ func (task *provisionerTask) setupToStartMachine(machine *apiprovisioner.Machine
 		possibleTools,
 	)
 	if err != nil {
-		task.removeMachineFromAZMap(machine)
-		task.setErrorStatus("cannot construct params for machine %q: %v", machine, err)
-		return nil, environs.StartInstanceParams{}, err
+		return environs.StartInstanceParams{}, errors.Annotatef(err, "cannot construct params for machine %q", machine)
 	}
 
-	return pInfo, startInstanceParams, nil
+	return startInstanceParams, nil
 }
 
 func (task *provisionerTask) startMachine(
 	machine *apiprovisioner.Machine,
-	provisioningInfo *params.ProvisioningInfo,
-	startInstanceParams environs.StartInstanceParams,
 	distributionGroupMachineIds []string,
 ) error {
+	v, err := machine.ModelAgentVersion()
+	if err != nil {
+		return err
+	}
+
+	startInstanceParams, err := task.setupToStartMachine(machine, v)
+	if err != nil {
+		return task.setErrorStatus("%v", machine, err)
```

Comment:
`task.removeMachineFromAZMap(machine)`, because setErrorStatus returns nil (should it, though?)

Reply:
No error from setupToStartMachine() should cause the provisioner to restart currently.  A machine isn't added to the AZMap until the end of the method via constructStartInstanceParams().  A machine isn't added to the AZMap until there is no failure cause left in constructStartInstanceParams().